Scaling Airbyte document. (#4630)
airbytehq/airbyte: docs/operator-guides/scaling-airbyte.md

should not present any scaling issues.

As a reference point, the typical Airbyte user has 5 - 20 connectors and 10 - 100 connections configured. Almost all of these connections are scheduled,
either hourly or daily, resulting in at most 100 concurrent jobs.

### What To Scale
[Workers](../understanding-airbyte/jobs.md) do all the heavy lifting within Airbyte. A worker is responsible for executing an Airbyte operations (e.g. Discover, Read, Sync etc),
and is created each time Airbyte need to do work. Thus, every job has a corresponding worker executing its work.

How a worker executes work depends on the Airbyte deployment. In the Docker deployment, an Airbyte worker spins up at least one docker container. In the Kubernetes
deployment, an Airbyte worker create at laest one Kubernetes pod. The created resource - Docker container or Kubernetes pod - does all the actual work.

Thus, scaling Airbyte is a matter of ensuring the Docker container/Kubernetes pod have sufficient resources to execute its work.

There are two resources to be aware of:
1) Memory
2) Disk space

##### Memory

##### Disk Space

> Airbyte used to use an on-disk queue resulting in significant disk space usage, especially for long-running syncs. Today Airbyte uses backpressure to try
> and read the minimal amount of logs required. Disk space is no longer a large concern as it was before.

Reviewer: ```suggestion
Airbyte uses backpressure to try and read the minimal amount of logs required. In the past, disk space was a large concern, but we've since deprecated the expensive on-disk queue approach.
```

> However disk space might be an issue for the following reasons:

Reviewer: ```suggestion
However, disk space might become an issue for the following reasons:
```

> 1) Long-running syncs can produce a fair amount of logs from the Docker agent and Airbyte on Docker deployments. Some work has been done to minimise
>    accidental logging so this should no longer be an acute problem. However, this is not guaranteed

Reviewer: ```suggestion
1) Long-running syncs can produce a fair amount of logs from the Docker agent and Airbyte on Docker deployments. Some work has been done to minimize accidental logging, so this should no longer be an acute problem, but is still an open issue.
```

   
2) Although Airyte connector images aren't massive, they aren't exactly small either. The typical connector image is ~300MB. An Airbyte deployment with
multiple connectors can easily use up to 10GBs of disk space.

> Because of these, we recommend allocating a minimum of 30GBs of disk space per node. Since storage is one of the cheaper resources, better to be safe than sorry!

Reviewer: ```suggestion
Because of this, we recommend allocating a minimum of 30GBs of disk space per node. Since storage is on the cheaper side, we'd recommend you being safe rather than sorry, so err on the side of overprovisioning.
```

This is because the Source worker reads up to 10,000 records in memory. This can present problems for database sources with tables that have large row sizes. e.g. a table with an average row size of 0.5MBs will require 0.5 * 10000 / 1000 = 5GBs of RAM. See [this issue](https://github.com/airbytehq/airbyte/issues/3439) for more information.

Our Java connectors currently follow Java's default behaviour with container memory and will only use up to 1/4 of the host's allocated memory. On Docker agent with 8GBs of RAM configured, a Java connector limits itself to 2Gbs of RAM and will see Out-of-Memory exceptions if this goes higher.

Note that all Source database connectors are Java connectors. This means that users currently need to over-specify memory resource for Java connectors.

On Docker, this can be solved by giving the Docker agent more memory. See [here](https://stackoverflow.com/questions/44533319/how-to-assign-more-memory-to-docker-container) for instructions. You might need to switch to a node with more memory. On Kubernetes, this can be solved by using a node type with more memory.

> 

Author: We should add section here on resource defaults. I'm thinking we can add this after @ChristopheDuong writes up documentation explaining how to specify our resource requirements.

Although this isn't complete yet, thinking it's alright to merge in as is since it's better than what we have today.

Author: also happy to do this myself. regardless I think we can tackle this in a follow up PR.
